netket/hilbert/index/constrained.py

```diff
@@ -36,79 +55,200 @@ def compute_constrained_to_bare_conversion_table(
     and is likely wrong.
     """
 
-    n_chunks = int(np.ceil(hilbert_index.n_states / chunk_size))
-    bare_number_chunks = []
-    for i in range(n_chunks):
-        id_start = chunk_size * i
-        id_end = np.minimum(chunk_size * (i + 1), hilbert_index.n_states)
-        ids = np.arange(id_start, id_end)
-
-        states = hilbert_index.numbers_to_states(ids)
-        is_constrained = constraint_fn(states)
-        (chunk_bare_number,) = np.nonzero(is_constrained)
-        bare_number_chunks.append(chunk_bare_number + id_start)
-
-    return np.concatenate(bare_number_chunks)
+    with jax.ensure_compile_time_eval():
+        n_chunks = int(np.ceil(hilbert_index.n_states / chunk_size))
+        bare_number_chunks = []
+        for i in range(n_chunks):
+            id_start = chunk_size * i
+            id_end = np.minimum(chunk_size * (i + 1), hilbert_index.n_states)
+            ids = jnp.arange(id_start, id_end, dtype=hilbert_index.index_dtype)
+            states = hilbert_index.numbers_to_states(ids)
+            is_constrained = constraint_fun(states)
+            (chunk_bare_number,) = jnp.nonzero(is_constrained)
+            bare_number_chunks.append(chunk_bare_number + id_start)
+        bare_numbers = jnp.concatenate(bare_number_chunks)
+    return bare_numbers
+
+
+class GenericConstrainedHilbertIndex(HilbertIndex):
+    _unconstrained_index: UnconstrainedHilbertIndex
+    _constraint_fun: Partial = struct.field(pytree_node=False)
+
```

Comment:
My issue with the previous code is that there were multiple wrappers of the function. The code becomes hard to read. Moreover, the intent is that we are hashing/recompiling every time this function changes, so I prefer this to be explicit instead of having a long series of wrappers constructed in the initialiser.

Assuming that the burden of ensuring hashability is on the user (which is the status quo at the moment), it is cleaner this way in my opinion.

Reply:
to put it differently, we don't gain anything by wrapping.